Examples/PSPDFCatalog/PSPDFCatalog/Catalog/Subclassing/CustomAnnotationProvider.cs

```diff
@@ -70,23 +72,44 @@ public CustomAnnotationProvider(PSPDFDocumentProvider documentProvider) : base(d
 					BoundingBox = new CGRect (100, pageInfo.Size.Height - 100, 32, 32),
 					Editable = false
 				};
-				if (!annotations.ContainsKey (pageIndex))
-					annotations.Add (pageIndex, new PSPDFAnnotation [] { noteAnnotation });
+				if (!annotationDict.ContainsKey (pageIndex))
+                    annotationDict.Add (pageIndex, new PSPDFAnnotation [] { noteAnnotation });
 			}
-			return annotations [pageIndex];
+			return annotationDict[pageIndex];
 		}
 
-		void PickColor (object o, EventArgs e)
+        public override PSPDFAnnotation[] AddAnnotations(PSPDFAnnotation[] annotations, NSDictionary<NSString, NSObject> options)
+        {
+            // Create an annotation list instead of an array because it's easier to work with
+            // We want to apply this to all annotation that are added (this only matters when you add 2 or more annotations at the same time via copy/paste etc)
+            foreach (var annotation in annotations)
+            {
+                // Need to add the already existing annotations
+                List<PSPDFAnnotation> existingAnnotations = annotationDict[annotation.PageIndex].ToList();
+                // Add all the other annotations once they're drawn
+                existingAnnotations.Add(annotation);
+                var annotationsAray = existingAnnotations.ToArray();
+                annotationDict[annotation.PageIndex] = annotationsAray;
+            }
+            return base.AddAnnotations(annotations, options);
+        }
+
+        void PickColor (object o, EventArgs e)
 		{
 			// Random Color
 			UIColor color = UIColor.FromRGBA (rnd.Next (0, 255), rnd.Next (0, 255), rnd.Next (0, 255), 1);
 			lock (this) {
-				foreach (var annotation in annotations) {
-					if (annotation.Value != null) {
-						annotation.Value [0].Color = color;
+				foreach (var annotations in annotationDict) {
+                    // We want every annotation to use the custom color
+                    foreach (var annotation in annotations.Value)
+                    {
+                        if (annotation != null)
+                        {
+                            annotation.Color = color;
 
-						ProviderDelegate.UpdateAnnotations (annotation.Value, true);
-					}
+                            ProviderDelegate.UpdateAnnotations(annotations.Value, true);
```

Comment:
This is called too often, since it is inside the annotation loop.
You could move this below the `foreach (var annotation in annotations.Value)` loop.